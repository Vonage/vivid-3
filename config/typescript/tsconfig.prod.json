{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "../../dist",
    "baseUrl": "../../src"
  },
  "exclude": [
    "../../src/**/*.spec.ts",
    "../../src/**/*.test.ts"
  ]
}
