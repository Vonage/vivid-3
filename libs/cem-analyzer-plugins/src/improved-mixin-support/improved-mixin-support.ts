import {
	extractMixinNodes,
	isMixin,
} from '@custom-elements-manifest/analyzer/src/utils/mixins.js';
import type {
	AnalyzePhaseParams,
	ModuleLinkPhaseParams,
	Plugin,
} from '@custom-elements-manifest/analyzer';

type ImprovedMixinSupportContext = {
	context: {
		mixinClassNames: Record<string, string>;
	};
};

/**
 * This plugin makes certain features (e.g. @attr/@event/@slot) work with mixins.
 * The analyzer has built in support for mixins, but it is flawed: the mixin declaration is created from the inner
 * class when the mixin is first encountered. That means that later plugins, like the @attr decorator plugin, will
 * not update the mixin. However, the inner class is still present in the moduleDoc and will be updated with the
 * attribute. The inner class is later dropped from the manifest as it is not exported. This plugin copies the
 * updated declarations from the inner class to the mixin before that happens.
 */
export const improvedMixinSupportPlugin = (): Plugin => ({
	name: 'improved-mixin-support',
	analyzePhase({
		ts,
		node,
		context,
	}: AnalyzePhaseParams & ImprovedMixinSupportContext) {
		switch (node.kind) {
			case ts.SyntaxKind.VariableStatement:
			case ts.SyntaxKind.FunctionDeclaration:
				if (isMixin(node)) {
					const { mixinFunction, mixinClass } = extractMixinNodes(node);
					const mixinName =
						mixinFunction?.name?.getText() ||
						mixinFunction?.parent?.name?.getText() ||
						mixinFunction?.declarationList?.declarations?.[0]?.name?.getText() ||
						'';
					// When we encounter a mixin, store the class name so that we can associate it with the mixin later
					context.mixinClassNames = context.mixinClassNames ?? {};
					context.mixinClassNames[mixinName] = mixinClass.name.getText();
				}
				break;
		}
	},
	moduleLinkPhase({
		moduleDoc,
		context,
	}: ModuleLinkPhaseParams & ImprovedMixinSupportContext) {
		for (const declaration of moduleDoc.declarations ?? []) {
			if (declaration.kind === 'mixin') {
				const mixinClassName = context.mixinClassNames[declaration.name];
				const mixinClass = moduleDoc.declarations!.find(
					(d) => d.kind === 'class' && d.name === mixinClassName
				) as any;
				Object.assign(declaration, {
					members: mixinClass.members,
					attributes: mixinClass.attributes,
					events: mixinClass.events,
					slots: mixinClass.slots,
				});
				if (mixinClass.vividTesting) {
					Object.assign(declaration, {
						vividTesting: mixinClass.vividTesting,
					});
				}
			}
		}
	},
});
