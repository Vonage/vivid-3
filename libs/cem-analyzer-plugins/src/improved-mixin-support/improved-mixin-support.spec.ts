import * as path from 'path';
import * as fs from 'fs';
import * as glob from 'glob';
import * as ts from '@custom-elements-manifest/analyzer/node_modules/typescript';
import { create } from '@custom-elements-manifest/analyzer/src/create.js';
import { improvedMixinSupportPlugin } from './improved-mixin-support';

const fixturesPath = path.join(__dirname, '__fixtures__');
const testCases = fs.readdirSync(fixturesPath);

test.each(testCases)(`Testcase %s`, async (testCase) => {
	const moduleFiles = glob.sync(
		path.join(fixturesPath, testCase, '/**/*.{ts,js}')
	);
	const modules = moduleFiles.map((moduleFile) =>
		ts.createSourceFile(
			'my-element.js',
			fs.readFileSync(moduleFile, 'utf-8'),
			ts.ScriptTarget.ES2015,
			true
		)
	);

	expect(
		create({ modules, plugins: [improvedMixinSupportPlugin()] })
	).toMatchSnapshot();
});
