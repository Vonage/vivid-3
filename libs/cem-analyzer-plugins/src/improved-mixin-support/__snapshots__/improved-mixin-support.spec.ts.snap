// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`improvedMixinSupportPlugin > should produce correct results for testcase default 1`] = `
{
  "modules": [
    {
      "declarations": [
        {
          "attributes": [
            {
              "name": "attribute",
            },
          ],
          "description": "",
          "events": [
            {
              "name": "mixin-event",
            },
          ],
          "kind": "mixin",
          "members": [
            {
              "kind": "field",
              "name": "attribute",
              "type": {
                "text": "string | undefined",
              },
            },
            {
              "kind": "field",
              "name": "property",
              "type": {
                "text": "string | undefined",
              },
            },
            {
              "kind": "method",
              "name": "method",
              "return": {
                "type": {
                  "text": "void",
                },
              },
            },
          ],
          "name": "MyMixin",
          "parameters": [
            {
              "name": "Base",
              "type": {
                "text": "T",
              },
            },
          ],
          "slots": [
            {
              "name": "mixin-slot",
            },
          ],
        },
      ],
      "exports": [
        {
          "declaration": {
            "module": "my-element.js",
            "name": "MyMixin",
          },
          "kind": "js",
          "name": "MyMixin",
        },
      ],
      "kind": "javascript-module",
      "path": "my-element.js",
    },
  ],
  "readme": "",
  "schemaVersion": "1.0.0",
}
`;
